pkg/ebpf/ingress_node_firewall_loader.go

```diff
@@ -166,113 +422,146 @@ func (infc *IngNodeFwController) makeIngressFwRulesMap(ingFirewallConfig ingress
+
+	return key, nil
+}
 
+// getStaleKeys goes through all existing rules and lists keys from the ebpfMap that do not match any
+// of the provided keys.
+func (infc *IngNodeFwController) getStaleKeys(desiredKeys []BpfLpmIpKeySt) ([]BpfLpmIpKeySt, error) {
 	objs := infc.objs
-	for _, ifaceName := range ifacesName {
-		ifaceName := ifaceName
-		// Look up the network interface by name.
-		iface, err := net.InterfaceByName(ifaceName)
-		if err != nil {
-			return ifMap, fmt.Errorf("lookup network iface %q: %s", ifaceName, err)
+
+	// Lookup all keys inside the map and find keys that are stale.
+	var staleKeys []BpfLpmIpKeySt
+	var key BpfLpmIpKeySt
+	var value BpfRulesValSt
+	iterator := objs.BpfMaps.IngressNodeFirewallTableMap.Iterate()
+	for iterator.Next(&key, &value) {
+		keyFound := false
+		for _, desiredKey := range desiredKeys {
+			if desiredKey.IngressIfindex == key.IngressIfindex &&
+				desiredKey.IpData == key.IpData &&
+				desiredKey.PrefixLen == key.PrefixLen {
+				keyFound = true
+				break
+			}
 		}
-		ifMap[ifaceName] = uint32(iface.Index)
-		if !isDelete {
-			// Attach the program.
-			l, err := link.AttachXDP(link.XDPOptions{
-				Program:   objs.IngressNodeFirewallProcess,
-				Interface: iface.Index,
+		if !keyFound {
+			staleKeys = append(staleKeys, BpfLpmIpKeySt{
+				PrefixLen:      key.PrefixLen,
+				IpData:         key.IpData,
+				IngressIfindex: key.IngressIfindex,
 			})
-			if err != nil {
-				if strings.Contains(err.Error(), xdpEBUSYErr) {
-					log.Printf("Interface %s is already attached", ifaceName)
-					return ifMap, nil
-				}
-				return ifMap, fmt.Errorf("could not attach XDP program: %s", err)
-			}
-			lPinDir := path.Join(infc.pinPath, ifaceName+"_link")
-			if err := l.Pin(lPinDir); err != nil {
-				return ifMap, fmt.Errorf("failed to pin link to pinDir %s: %s", lPinDir, err)
-			}
-			infc.links[ifaceName] = l
-			log.Printf("Attached IngressNode Firewall program to iface %q (index %d)", iface.Name, iface.Index)
-		} else {
-			log.Printf("Unattaching IngressNode Firewall program from iface %q (index %d)", iface.Name, iface.Index)
-			if err := infc.cleanup(ifaceName); err != nil {
-				return ifMap, err
-			}
 		}
 	}
-	return ifMap, nil
+	err := iterator.Err()
+	if err != nil {
+		return staleKeys, err
+	}
+
+	return staleKeys, nil
 }
 
-// cleanup will delete interface's eBPF objects.
-func (infc *IngNodeFwController) cleanup(ifName string) error {
-	l, ok := infc.links[ifName]
-	if !ok {
-		return fmt.Errorf("failed to find Link object for interface %s", ifName)
+// getStaleInterfaceKeys returns the keys for all rules that belong to stale interfaces, meaning interfaces
+// that are not attached any more.
+func (infc *IngNodeFwController) getStaleInterfaceKeys() ([]BpfLpmIpKeySt, error) {
+	objs := infc.objs
+
+	// Looup all valid interfaces IDs.
+	var validInterfaceIDs []uint32
+	for interfaceName := range infc.links {
+		iface, err := net.InterfaceByName(interfaceName)
```

Comment:
use the new interfaces method here as well

Reply:
done